Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,150 @@
+                add(b.finalize())
+            }
+        }
+    }
+
+    /// Adds an individual program to the corpus
+    public func add(_ program: Program) {
+        if program.size > 0 {
+            let execution = fuzzer.execute(program)
+            guard execution.outcome == .succeeded else { return }
+            if let aspects = fuzzer.evaluator.evaluate(execution) {
+                add(program, aspects)
+            }
+        }
+    }
+
+    /// Adds multiple programs to the corpus.
+    public func add(_ programs: [Program]) {
+        logger.info("Markov Corpus import of \(programs.count) programs")
+        for (index, prog) in programs.enumerated() {
+            if index % 500 == 0 {
+                logger.info("Markov Corpus import at \(index) of \(programs.count)")
+            }
+            add(prog)
+        }
+    }
+
+    public func add(_ program: Program, _ aspects: ProgramAspects) {
+        if program.size > 0 {
+            self.allIncludedPrograms.append(program)
+            let edges = aspects.toEdges()
+            for e in edges {
+                self.edgeMap[e] = program
+            }
+        }
+    }
+
+    // Switch evenly between programs in the current queue and all programs available to the corpus
+    public func randomElement() -> Program {
+        assert(self.size > 0)
+        if let prog = self.programExecutionQueue.randomElement(), probability(0.5) {
+            assert(prog.size != 0)
+            return prog
+        }
+        let prog = self.allIncludedPrograms.randomElement()!
+        assert(prog.size != 0)
+        return prog
+    }
+
+    public func getNextSeed() -> (seed: Program, energy: UInt64) {
+        self.totalExecs += 1
+        assert(self.size > 0)
+        // Only do computationally expensive work choosing the next program when there is a solid
+        // baseline of execution data
+        if self.totalExecs > 500 {
+            // Check if more programs are needed
+            if self.programExecutionQueue.count == 0 {
+                self.regenProgramList()
+            }
+            assert(self.programExecutionQueue.count > 0)
+            let prog = self.programExecutionQueue.popLast()!
+            return (prog, self.energyBase())
+        } else {
+            let element = self.allIncludedPrograms.randomElement()!
+            return (element, 1)
+        }
+    }
+
+    // Regenerates the internal program list, with dropoff to ensure some diversity among instances
+    // Ends up with ~ 1/32th of the Corpus
+    private func regenProgramList() {
+        assert(self.programExecutionQueue.count == 0)
+        // TODO: Hook things up so that configurable constant "numConsecutiveMutations" is used rather than 5
+        let edges = self.fuzzer.evaluator.smallestEdges(desiredEdgeCount: UInt64(self.size/8), expectedRounds: self.energyBase() * 5)!.toEdges()
+        for e in edges {
+            if let prog = self.edgeMap[e], probability(0.25) {
```

Comment:
Is it expected that `edgeMap` has a program for every edge that has been discovered? If so, maybe add a `logger.warning` if that isn't the case?
I suspect it's currently not the case on workers after the initial synchronization. I'm wondering if we'd need a way to retrieve all edges triggered by a given sample (`CoverageEvaluator.evaluate` only returns the newly discovered edges) to then recompute this `edgeMap` on the workers (and maybe even periodically on all instances?). That should be fairly easy to do since it's just the content of the shared-memory coverage bitmap I think. WDYT?

Reply:
I do expect that `edgeMap` has a program for each found edge, and adding a warning makes sense. I believe that due to the way `add` is implemented, all programs should have been executed and properly tracked in the edge map before generating a new list of programs to fuzz. I'll add the warning, and run the fuzzer for a bit to see what pops up.

One of the things I've been thinking about is how to handle found edges that are only hit due to non-determinism, and rerunning the sample likely will not hit that edge (this is the reason for the aggressive aging of samples this corpus selects, before actually running them, to ensure the instance doesn't get stuck on a sample that rarely hits the edge that caused it to be added). I've noticed that when I import the corpus upon resuming a session, a significant number of programs in the original corpus drop off. I believe this could be due to non-determinism in execution, or some programs in the corpus being subsets of others. What do you think about running a periodic re-import of all samples, regardless of corpus implementation? `BasicCorpus` would ensure each one still hits a new edge consistently, and `MarkovCorpus` would additionally ensure it has an accurate `edgeMap`

Reply:
Interesting idea! Yeah afaict non-determinism is definitely a problem. I had also thought of implementing a `--deterministic` mode where each sample is run at least a second time before being included in the corpus. But I think your idea is better. That would probably be somewhat similar to the corpus culling feature that AFL uses? I think that also removes samples from the corpus whose edges are covered by other (smaller) samples. Maybe this implementation could just re-execute the samples starting from the latest, so that old samples that are fully "subsumed" by newer ones would automatically drop out. I'm happy to give such a mechanism a try, either as part of this PR or as a separate one!

Reply:
I think it makes sense as part of [issue 26](https://github.com/googleprojectzero/fuzzilli/issues/26). It seems that the `MarkovCorpus` has some protection against non-deterministic samples imported instances, as my long fuzzing sessions result in the `BasicCorpus` instances having more samples than the `MarkovCorpus` instances, with a number of samples falling off during resuming the session. I'll add some more comments to that issue.